modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java

```diff
@@ -782,4 +803,120 @@ private class HeartbeatTask extends TimerTask {
             }
         }
     }
+
+    /**
+     * Timeout object worker.
+     */
+    private static class TimeoutWorker extends IgniteWorker {
+        /** Worker sleep interval. */
+        private final long sleepInterval = getLong("IGNITE_TIMEOUT_WORKER_SLEEP_INTERVAL", 500);
+
+        /** Active operation. */
+        public final ConcurrentMap<Long, TimeoutObject> requestsMap;
+
+        /**
+         * Constructor.
+         *
+         * @param log Logger.
+         * @param igniteInstanceName Name of the Ignite instance this runnable is used in.
+         * @param name Worker name. Note that in general thread name and worker (runnable) name are two different things. The same
+         *         worker can be executed by multiple threads and therefore for logging and debugging purposes we separate the two.
+         * @param requestsMap Active operations.
+         */
+        public TimeoutWorker(
+                IgniteLogger log,
+                String igniteInstanceName,
+                String name,
+                ConcurrentMap<Long, TimeoutObject> requestsMap
+        ) {
+            super(log, igniteInstanceName, name, null);
+
+            this.requestsMap = requestsMap;
+        }
+
+        @Override
+        protected void body() {
+            try {
+                TimeoutObject timeoutObject;
+
+                while (!isCancelled()) {
+                    long now = coarseCurrentTimeMillis();
+
+                    for (Entry<Long, TimeoutObject> entry : new HashMap<>(requestsMap).entrySet()) {
+                        updateHeartbeat();
+
+                        timeoutObject = entry.getValue();
+
+                        assert timeoutObject != null : "Unexpected null on the timeout queue.";
+
+                        if (timeoutObject.getEndTime() > 0 && now > timeoutObject.getEndTime()) {
+                            CompletableFuture<?> fut = timeoutObject.getFuture();
+
+                            if (!fut.isDone()) {
+                                fut.completeExceptionally(new TimeoutException());
+                            }
+
+                            // Client-facing future will fail with a timeout, but internal ClientRequestFuture will stay in the map -
+                            // otherwise we'll fail with "protocol breakdown" error when a late response arrives from the server.
+                            // requestsMap.remove(entry.getKey(), timeoutObject)
```

Comment:
Why is this line commented out?

Reply:
Because of that comment above,. the protocol has to get a response in any case. It is a main difference between timeout worker for Ignite node and the client. That is also a reason why I do not use one class in both cases.

Reply:
Then why do you leave the commented-out line? You explained why you don't remove from the map, isn't this enough?

Reply:
I have rewritten this place.